tools/golang/cmd/encrypt/script.go

```diff
@@ -0,0 +1,120 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package main
+
+import (
+	"errors"
+	"flag"
+	tinkpb "github.com/google/tink/proto/tink_go_proto"
+	"github.com/subscriptions-project/swg-js/tools/golang/encryption"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+)
+
+/* Script to encrypt documents for the SwG Encryption Project.
+*
+* Golang script which takes in an input HTML document and encrypts
+* all content within <section subscriptions-section="content" encrypted>
+* tags using AES-GCM. The key used to encrypt the content is added
+* to the output document's head inside of a
+* <script cryptokeys type="application/json"> element. The encrypted
+* document is outputted to the output_file path given as a flag.
+*
+* Example Usage:
+* go run swg-js/tools/golang/cmd/encrypt/script.go \
+*	--input_html_file=../tmp/sample-encryption.html \
+*	--output_file=../tmp/sample-encryption-out.html \
+*	--google_public_key_url=https://news.google.com/swg/encryption/keys/{dev|prod}/tink/public_key \
+*	--access_requirement=norcal.com:premium \
+*	--publisher_public_key_url=nytimes.com,www.nytimes.com/scs/publickey \
+*	--publisher_public_key_url=wp.com,www.wp.com/scs/publickey
+ */
+type mapFlags map[string]string
+
+const googleDevPublicKeyUrl string = "https://news.google.com/swg/encryption/keys/dev/tink/public_key"
+
+func (m *mapFlags) String() string {
+	var strs []string
+	for key, val := range *m {
+		strs = append(strs, key, ";", val)
+	}
+	return strings.Join(strs, "\n")
+}
+func (m *mapFlags) Set(value string) error {
+	s := strings.Split(value, ",")
+	if len(s) != 2 {
+		return errors.New("Malformatted value inserted")
+	}
+	(*m)[s[0]] = s[1]
+	return nil
+}
+
+func main() {
+	// Input flags.
+	inputHtmlFile := flag.String("input_html_file", "", "Input HTML file to encrypt.")
+	outFile := flag.String("output_file", "", "Output path to write encrypted HTML file.")
+	accessRequirement := flag.String("access_requirement", "", "The access requirement we grant upon decryption.")
+	mf := make(mapFlags)
+	flag.Var(&mf, "encryption_key_url", `Strings in the form of '<domain-name>,<url>', where url is 
+										 link to the hosted public key that we use to encrypt the 
+										 document key.`)
+	flag.Parse()
+	if *inputHtmlFile == "" {
+		log.Fatal("Missing flag: input_html_file")
+	}
+	if *outFile == "" {
+		log.Fatal("Missing flag: output_file")
+	}
+	if *accessRequirement == "" {
+		log.Fatal("Missing flag: access_requirement")
+	}
+	// Read the input HTML file.
+	b, err := ioutil.ReadFile(*inputHtmlFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// Retrieve all public keys from the input URLs.
+	pubKeys := make(map[string]tinkpb.Keyset)
+	var pubKey tinkpb.Keyset
+	for domain, url := range mf {
+		pubKey, err = encryption.RetrieveTinkPublicKey(url)
+		if err != nil {
+			log.Fatal(err)
+		}
+		pubKeys[domain] = pubKey
+	}
+	if _, ok := pubKeys["google.com"]; !ok {
```

Comment:
I agree the duplication doesn't look nice, but I added it because I don't want to override the user's chosen value for google.com, especially if they use the prod URL.

Reply:
Can you check if they specify google.com before the for loop?

```
if _, ok := mf["google.com"]; !ok {
  mf["google.com"] = googleDevPublicKeyUrl
}
for domain, url := range mf {
  ...
}
```

Reply:
Nice!